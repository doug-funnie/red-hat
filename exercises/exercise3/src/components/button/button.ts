import glamorous from 'glamorous';
import { lighten } from 'polished';
import { ButtonHTMLAttributes } from 'react';
import { ThemedProps } from 'styles/theme';

type Props = Pick<ButtonHTMLAttributes<HTMLButtonElement>,
  'onClick'
  | 'disabled'
  | 'type'
>;

export const Button = glamorous.button<ThemedProps<Props>>(({ theme, disabled }) => {
  const bg = theme.colors.red;
  const activeBg = lighten(0.2, bg);

  return {
    border: 'none',
    borderRadius: 4,
    fontWeight: 700,
    height: 32,
    lineHeight: '32px',
    outline: 'none',
    padding: '0 10px',
    color: theme.colors.white,
    backgroundColor: bg,
    opacity: disabled ? 0.6 : 1,
    cursor: disabled ? 'not-allowed' : 'cursor',
    ':hover': {
      backgroundColor: activeBg,
    },
    ':active': {
      backgroundColor: activeBg,
    },
    ':focus': {
      backgroundColor: activeBg,
    },
  };
});

Button.defaultProps = {
  type: 'button',
};
