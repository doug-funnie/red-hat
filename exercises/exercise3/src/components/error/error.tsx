import { Div } from 'glamorous';
import * as React from 'react';
import { Button } from '../button';
import { H2, H3 } from '../text';

export interface Props {
  title?: string;
  message?: string;
  icon?: string;
  onRetry?(): void;
}

const defaultProps: Partial<Props> = {
  icon: 'ðŸ˜’',
  title: 'Oops!',
  message: 'Something went wrong.',
};

export const Error: React.SFC<Props> = ({
  title,
  message,
  icon,
  onRetry,
}) => (
  <Div display='flex' flexDirection='column' alignItems='center' justifyContent='center'>
    <Div fontSize={75}>{icon}</Div>
    <Div marginBottom={15} textAlign='center'>
      <H2>{title}</H2>
      <H3>{message}</H3>
    </Div>
    {!!onRetry && (
      <Button onClick={onRetry}>
        Give it another go
      </Button>
    )}
  </Div>
);

Error.defaultProps = defaultProps;
