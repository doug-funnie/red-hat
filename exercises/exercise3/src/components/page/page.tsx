import glamorous, { Main } from 'glamorous';
import * as React from 'react';
import { Error } from '../error';
import { Header, height as headerHeight } from '../header';

export interface Props {
  title?: string;
  children?: React.ReactNode;
  error?: any;
  onBackButtonPress?(): void;
  onRetry?(): void;
}

const Wrapper = glamorous.div({
  margin: '0 auto',
  maxWidth: 800,
  paddingBottom: 25,
  paddingTop: 15,
  width: '100%',
});

export const Page: React.SFC<Props> = ({
  children,
  title,
  onBackButtonPress,
  error,
  onRetry,
}) => (
  <>
    <Header onBackButtonPress={onBackButtonPress} title={error ? 'Error' : title} />
    <Main marginTop={headerHeight}>
      <Wrapper>
        {!!error ? <Error onRetry={onRetry} /> : children}
      </Wrapper>
    </Main>
  </>
);
