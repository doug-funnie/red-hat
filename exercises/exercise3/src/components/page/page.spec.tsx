import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Button } from '../button';
import { Error } from '../error';
import { Page, Props } from './page';

const content = <div>content</div>;

const props: Props = {
  title: 'Title',
  children: content,
};

it('renders children if no error is present', () => {
  const view = shallow(<Page {...props} />);
  expect(toJson(view)).toMatchSnapshot();
});

it('renders error if error is present', () => {
  const view = shallow(<Page {...props} error='error' />);
  expect(view.find(Error).length).toBe(1);
});
