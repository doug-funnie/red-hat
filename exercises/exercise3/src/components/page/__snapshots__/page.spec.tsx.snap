// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders children if no error is present 1`] = `
<React.Fragment>
  <Component
    title="Title"
  />
  <glamorous.Main
    marginTop={60}
  >
    <glamorous(div)>
      <div>
        content
      </div>
    </glamorous(div)>
  </glamorous.Main>
</React.Fragment>
`;
